src/Build/Instance/ProjectInstance.cs

```diff
@@ -2942,11 +2942,9 @@ private void CreateEvaluatedIncludeSnapshotIfRequested(bool keepEvaluationCache,
                 if (item.DirectMetadata != null)
                 {
                     directMetadata = new CopyOnWritePropertyDictionary<ProjectMetadataInstance>();
-                    foreach (ProjectMetadata directMetadatum in item.DirectMetadata)
-                    {
-                        ProjectMetadataInstance directMetadatumInstance = new ProjectMetadataInstance(directMetadatum);
-                        directMetadata.Set(directMetadatumInstance);
-                    }
+
+                    IEnumerable<ProjectMetadataInstance> projectMetadataInstances = item.DirectMetadata.Select(directMetadatum => new ProjectMetadataInstance(directMetadatum));
```

Comment:
Is Select and delegate here as efficient as a plain loop?

The core libraries would not use this, I think (@stephentoub?)

Reply:
> Is Select and delegate here as efficient as a plain loop?

It depends on several factors, including what DirectMetadata is.

> The core libraries would not use this, I think (@stephentoub?)

It also depends :) but generally we avoid use of `IEnumerable<T>` manipulation on hot paths.

Reply:
Purpose of this changes was to eliminate generation of `ImmutableDictionary` in a loop which has very very bad  CPU and memory allocation consequences. 
To do that I decided to use prior existing `internal void ImportProperties(IEnumerable<T> other)` method which calls  `public ImmutableDictionary<TKey, TValue> SetItems(IEnumerable<KeyValuePair<TKey, TValue>> items)` internally.
I do not see how to avoid `IEnumerable<T>` here, unless we would change `CopyOnWritePropertyDictionary` API to somehow expose immutable builder. I don't think this path is that hot though.